Examples/KlaviyoSwiftExamples/SPMExample/NotificationServiceExtension/NotificationService.swift

```diff
@@ -0,0 +1,168 @@
+                if let imageAttachment = imageAttachment {
+                    bestAttemptContent.attachments = [imageAttachment]
+                }
+
+                /// Call the content handler to deliver the push notification.
+                /// NOTE: if this doesn't happen before the stipulated time then the completion handler is called with
+                /// the bestAttemptContent from the `serviceExtensionTimeWillExpire` function
+                contentHandler(bestAttemptContent)
+            }
+        }
+    }
+
+    override func serviceExtensionTimeWillExpire() {
+        /// Called just before the extension will be terminated by the system.
+        /// Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
+        if let contentHandler = contentHandler,
+           let bestAttemptContent = bestAttemptContent {
+            contentHandler(bestAttemptContent)
+        }
+    }
+}
+
+// MARK: supporting methods to notification service
+
+extension NotificationService {
+    private func getMedia(
+        for urlString: String,
+        completion: @escaping (UIImage?) -> Void) {
+        guard let url = URL(string: urlString) else {
+            completion(nil)
+            return
+        }
+        downloadImage(forURL: url) { result in
+            switch result {
+            case let .success(image):
+                completion(image)
+            case let .failure(error):
+                print("received error [\(error.localizedDescription)]")
+                completion(nil)
+            }
+        }
+    }
+
+    private func saveImageAttachment(
+        image: UIImage,
+        forIdentifier identifier: String) -> URL? {
+        /// Obtain a reference to the temp file directory.
+        let tempDirectory = URL(fileURLWithPath: NSTemporaryDirectory())
+
+        /// Using the temp file directory, create a directory URL using a unique string.
+        let directoryPath = tempDirectory.appendingPathComponent(
+            ProcessInfo.processInfo.globallyUniqueString,
+            isDirectory: true)
+
+        do {
+            /// The FileManager is responsible for creating the actual file to store the data.
+            /// Call `createDirectory(at:winthIntermediateDirectories:attributes:)` to create an empty directory
+            try FileManager.default.createDirectory(
+                at: directoryPath,
+                withIntermediateDirectories: true,
+                attributes: nil)
+
+            /// Create a file URL based on the image identifier.
+            let fileURL = directoryPath.appendingPathComponent(identifier)
+
+            /// Create a Data object from the image.
+            guard let imageData = image.pngData() else {
+                return nil
+            }
+
+            /// Attempt to write the file to disk.
+            try imageData.write(to: fileURL)
```

Comment:
I feel like you mentioned this somewhere, are we limited in storage space, should we remove old images somehow?

Reply:
Yeah. I did try and look this up but the documentation from apple on this is very vague and not much on SO either. I'm going to create a task for us to look into this but for now I'm going to go ahead with merging this code.